sonic_platform_base/sonic_xcvr/api/public/cmis.py

```diff
@@ -1745,4 +1765,246 @@ def get_transceiver_loopback(self):
+        """
+        state = {
+            'module_state': self.get_module_state(),
+            'config_state': self.get_config_datapath_hostlane_status(),
+            'datapath_state': self.get_datapath_state()
+        }
+        return state
+
+    def get_cmis_application_selected(self, host_lane):
+        """
+        Get the CMIS selected application code of a host lane
+
+        Args:
+            host_lane:
+                Integer, the lane id on the host/system side
+
+        Returns:
+            Integer, the transceiver-specific application code
+        """
+        ap_code = 0
+        if host_lane in range(self.NUM_CHANNELS) and not self.is_flat_memory():
+            field = "{}_{}_{}".format(consts.STAGED_CTRL_APSEL_FIELD, 0, host_lane + 1)
+            ap_code = self.xcvr_eeprom.read(field) >> 4
+
+        return (ap_code & 0xf)
+
+    def get_cmis_application_matched(self, host_speed, host_lanemask):
+        """
+        Get the CMIS application code that matches the specified host side configurations
+
+        Args:
+            host_speed:
+                Integer, the port speed of the host interface
+            host_lanemask:
+                Integer, a bitmask of the lanes on the host side
+                e.g. 0x5 for lane 0 and lane 2.
+
+        Returns:
+            Integer, the transceiver-specific application code
+        """
+        if host_speed == 0 or host_lanemask == 0:
+            return 0
+
+        host_lane_count = 0
+        for lane in range(self.NUM_CHANNELS):
+            if (1 << lane) & host_lanemask == 0:
+                continue
+            host_lane_count += 1
+
+        appl_code = 0
+        appl_dict = self.xcvr_eeprom.read(consts.APPLS_ADVT_FIELD)
+        if appl_dict is None or appl_dict.strip() in ["None", "{}", ""]:
+            return 0
+        appl_dict = ast.literal_eval(appl_dict)
+        for c in appl_dict.keys():
+            d = appl_dict[c]
+            if d.get('host_lane_count') != host_lane_count:
+                continue
+            if self.get_host_speed(d.get('host_electrical_interface_id')) != host_speed:
+                continue
+            appl_code = c
+            break
+
+        return (appl_code & 0xf)
+
+    def has_cmis_application_update(self, host_speed, host_lanemask):
+        """
+        Check for CMIS application update and retrieve the new application code
+
+        Args:
+            host_speed:
+                Integer, the port speed of the host interface
+            host_lanemask:
+                Integer, a bitmask of the lanes on the host side
+                e.g. 0x5 for lane 0 and lane 2.
+
+        Returns:
+            (has_update, new_appl)
+        """
+        if host_speed == 0 or host_lanemask == 0 or self.is_flat_memory():
+            return (False, 1)
+
+        app_new = self.get_cmis_application_matched(host_speed, host_lanemask)
+        if app_new != 1 or host_lanemask != (1 << self.NUM_CHANNELS) - 1:
```

Comment:
why do you think the default application should ALWAYS support 8 lanes - this is not realistic assumption

Reply:
Done, this additional checker is removed